### DISCUSSION TITLE
QPDF "pages epic"

### DISCUSSION THREAD

**Discussion by jberkenbilt** [qpdf/qpdf#1104] [Ideas]
I am ready to dig into the implementation of the _pages epic_. This work is related to all the issues I've been tagging with the `pages` label for the last several years (!). It consists of two broad categories of work:
* Better handling of document-level data with page split/merge operations to avoid things like breaking hyperlinks, layers, and outlines (bookmarks) when merging and splitting files
* Offering more flexible ways of assembling pages including forming groups based on document-level features (outlines, threads, tags, etc.), doing n-up or other page-composition options, controlling placement of pages with overlay/underlay, etc.

I have written a draft of what I plan to do. That can be found in the [TODO-pages.md](https://github.com/qpdf/qpdf/blob/main/TODO-pages.md) file at the top of the repository.

**My preferred way to accept comments is via discussion in this issue as opposed to something like a pull request against that file.**

This design is taking the shape of how I always do designs. Early steps are clear, and the farther out we go, the fuzzier it gets. There are some details I've left vague because I haven't yet figured them out. As I get to the point of being ready to implement some functionality, that part of the design will be more strongly solidified.

I don't know how long it will take to implement everything here. Some parts of this may never get done. My design allows for changes to be introduced incrementally. Rather than repeating the details here, I'll refer you to the document. I'm looking forward to this experiment -- mostly I have designed features either in a total vacuum or in consultation with one or two people. This is the first time I have openly shared my design this early in hopes of getting more community input. That said, I'm prepared to plow through this if needed!

**Comment by jberkenbilt** [comment]
@jbarlow83 @m-holger Letting you know specifically about the pages epic discussion issue. Your input would be gratefully received. @jbarlow83 -- in particular, I'm interested in your feedback on my idea of having new `QPDFAssembler` and `QPDFSplitter` classes that provide a higher-level interface, but not as high as `QPDFJob` for doing page manipulation. It would always be possible for people to continue to call `addPage`, but using `QPDFAssembler` would allow people to manipulate pages in multiple PDFs and letting the library deal with maintaining functionality offered by document-level constructs. I am hoping to have something that's a little more natural from a programming standpoint than `QPDFJob`, which is mainly intended to encapsulate work performed by the qpdf CLI.

As always, my intention is to keep existing code working. I don't know yet whether this will all fit into qpdf 11 or whether I will need to break ABI. At this time, nothing in the design requires me to break ABI, but it's likely that I will decide to open an ABI window at some point, particularly so we can deal with things like `std::string_view` or doing things to reduce the number of copies. @m-holger Of course your input is welcome wherever you wish to provide it, but I'm hoping you'll catch me if I start creating interfaces that result in needless copies or are overly cumbersome. I'm more conscious of this than in the past.

For everyone else, all input is, of course, welcome. This could include use cases, usability concerns, syntax suggestions, or anything else.

My intention is to update this issue from time to time as I make progress.

**Comment by jbarlow83** [comment]
Looks very ambitious!

**Interface**

Generally the QPDFAssembler/Splitter API division sounds good to me and would be easier to manage than QPDFJob. 

On my end I might end up creating these objects as needed but maybe not exposing bindings to them directly. It depends what exactly they end up looking like.

For example currently in pikepdf, a statement like `pdf.pages[2:4] = other_pdf.pages[0:2]`. The `other_pdf.pages[0:2]` creates a list of `pikepdf.Page/QPDFPageObjectHelper` then assignment is implementing by inserting pages from that view into `pdf` and deleting the old pages. I'd probably use QPDFSplitter to create the slice instead to retain all of the information, and then QPDFAssembler on the other side to merge the sliced pages. For that to work I'd need QPDFAssembler to update the primary input QPDF rather than returning a new QPDF, since the existing pikepdf API is expressed in terms of modifying the state of the existing primary PDF. 

On the other hand this interface might break down for more complex operations because of its need for immediate evaluation of all intermediate steps. If QPDFAssembler/QPDFSplitter are more along the lines of describing a lazily evaluated plan to transform a  PDF, then the list-like interface is probably the wrong model. I'd need something more like a declarative interface for how one wants the pages organized. (Come to think of it, that would probably be better.)

**Page labels?**

Might need to way to specify whether pages given on the command line are page numbers or page labels that happen to be numeric. (If that isn't already present?)

**StructTreeRoot**

Having StructTreeRoot copying would be great. Accessible PDFs are becoming more important in many fields and being able to transform them with PDF software confidently is also important. But it's complex.

I recommend using PDF/UA documents as some of your test inputs, so you can use [PAC 2021](https://pdfua.foundation/en/pdf-accessibility-checker-pac/) to validate correct handling. [pdfbrewer](https://www.cib.de/en/pdfbrewer/) can save as PDF/UA. PAC 2021 is good at checking StructTreeRoot for problems, while Acrobat mostly whines "there's something wrong" and Preflight can point in the general direction of certain issues but doesn't always know what's wrong. Unsurprisingly there are a lot of malformed StructTreeRoots out there.

StructTreeRoot have a complex interaction with Form XObjects. To highlight one relevant and messy detail: if you're going to turn the page into a Form XObject for those transformations, you need to update the StructTreeRoot to point to the content stream inside the Form XObject instead of just the destination page, since every content stream has its own set of MCID numbers. This is discussed in section 14.7.5.2. Now for another twist, a /K can is array of one of four types of reference, and Form XObjects can contain other Form XObjects.

**Low level**

The method chaining/fluent interface of QPDFJob by raw pointers is awkward to deal with in pybind11. I ended implementing only the Job JSON and CLI interfaces. The issue being that when pybind11 sees a raw pointer, it wants to take ownership, and wrap it in a holder and manage it. I know of one workaround I haven't tried yet, but pybind11 has open unsolved issues in that general area with complex Python/C++ reference cycles. Returning a reference or lightweight copyable object like QPDFObjectHandle works best, I've found. 

>Allow creation of blank pages as an additional input source

I think it's fine to implement this as creating a new empty QPDF and attaching a blank page to it, rather than treating blank pages as a special input source. It wouldn't hurt to have a little API syntactic sugar around that process, though.

**Reply by jberkenbilt** [reply]
These are great responses. Thanks for taking the time to read and comment. It is much appreciated!

> On my end I might end up creating these objects as needed but maybe not exposing bindings to them directly. It depends what exactly they end up looking like. ... For that to work I'd need QPDFAssembler to update the primary input QPDF rather than returning a new QPDF, since the existing pikepdf API is expressed in terms of modifying the state of the existing primary PDF.

Yes, I think we're thinking the same way.

I would think of `QPDFAssember` as being something you would most likely use behind the scenes. When someone manipulates pages with arrays like that, you could call `QPDFAssember` instead of other methods, and that would delegate all the document-level stuff to `QPDFAssember`. I view pikepdf as having its own higher-level interfaces for qpdf that are more Pythonic. Of course, you could also expose `QPDFAssember` if you felt like it, and I'd like to not preclude that by some aspect of my design. I occasionally use pikepdf, and I find its interfaces to be extremely natural. It is one of the best Python packages I've used.

Regarding the original `QPDF`, `QPDFAssember` would work exactly the way `QPDFJob` works today. There is a "primary input", and all operations are modifications to that. It is my full intention to allow any `QPDF` object to be a primary input, so, yes, it would make changes to the original `QPDF`. I updated the doc to state that, if a `QPDF` is the primary input, the same `QPDF` will be returned. What `QPDFJob` does today is to track all operations and ensure that the first occurrence of a page from the primary input as it appears in the output, even if the order has changed, is the exact same page object with the same ID. Internally, qpdf clears the pages tree up front and then re-inserts pages. When it's completely done, it deletes any unreferenced pages. I think this model will work well with what you're trying to achieve. It also preserves document-level functionality for the primary input even for things I haven't gotten to yet.

> I'd need something more like a declarative interface for how one wants the pages organized. (Come to think of it, that would probably be better.)

Yes, this is how I see it. `QPDFAssembler` (and currently `QPDFJob`) sequences things on its own as described in the document. That way it can do all the calculations and move things around in the most efficient way. For example, overlay and underlay are always done after page ordering, and when I add composition, that will always be done last. If someone wanted to do things in a different order, they'd have to create intermediate results. From the CLI, that would be to create intermediate files, but from C++ code, you could take the QPDF from one `QPDFAssembler` and use it as an input to another. It should even work to stack `QPDFAssembers` with the same `QPDF`. This is not a common pattern right now because it's hard to set up, but I have test cases to make sure it works. It gets quite tricky when you do things like copying a stream that has a stream data provider, and I can't promise there aren't some bugs, but I think it's quite solid overall.

> Might need to way to specify whether pages given on the command line are page numbers or page labels that happen to be numeric. (If that isn't already present?)

Right now, it's always sequential page numbers from 1. It would be nice to have something based on page labels. It looks like I had that and accidentally dropped it. I just went through the last pre-rework version of my TODO document and recovered a few cases I had previous thought of including selecting pages based on labels. I haven't thought of a syntax for that, but it would probably be something like `--labels=....` rather than `--range=...`. Once the new framework is in place, adding something like that would be quite easy.

> StructTreeRoot

Your comments are noted. I haven't really studied this part of the spec carefully yet. It appears to be the most complex bit, so it probably won't be first. I'll make a note to come back to your comments.

> The method chaining/fluent interface of QPDFJob by raw pointers is awkward to deal with in pybind11. .... Returning a reference or lightweight copyable object like QPDFObjectHandle works best, I've found.

I didn't know this about pybind -- this is very helpful and is one of the reasons I'm interested in having your early input. I think using reference should be fine and will be efficient and flexible. I'll try that first. Would `std::shared_ptr` work? I know there are ways to create a shared_ptr to `this` that weren't there in earlier C++ versions, but I'll have to refresh myself. I only use C++ in qpdf now...most of the code I write in my day job is in Go, and if it's not in Go, it's probably in Python. It's hard to keep up with all the changes in the C++ standard library. Anyway, if raw pointers work, then references will work too since I think there are no cases in a fluent interface where you would ever return a null pointer.

> I think it's fine to implement this as creating a new empty QPDF and attaching a blank page to it, rather than treating blank pages as a special input source. It wouldn't hurt to have a little API syntactic sugar around that process, though.

By "input source", I just mean a thing inside `--pages ... --`. My intention is just create a page object with a media box and no content and add it to the pages tree, so this is about as light-weight as it gets.

**Reply by jbarlow83** [reply]
Thanks, that helps me see what you have in mind.

For fluent interfaces, returning an object reference is best, i.e. `QPDFAssembler& QPDFAssembler::setFribbleBit(bool bit);`

**Reply by m-holger** [reply]
I agree shared pointers should be avoided. Where reference don't work, unique pointers would be preferable.

**Comment by m-holger** [comment]
I haven't had time to read and consider carefully, so I am somewhat shooting from the hip here. On a first read the only two items from my personal shopping list that are missing are:

- allow modifications to be specified inside the `--pages ... --` section

  i.e instead of `--pages file1 1-3 file 2 z-5 -- --rotate=90:3,27`, accept `--pages file1 1-2 file1 --rotate=90 3 file2 z-2 file 2 --rotate=90 1 --`.
  (I'm not clear from TODO-pages whether this is already intended.)
 
- provide a partial configuration object to provide a less verbose way to specify multiple options 
 
  Lets call it style for the discussion.  What I have in mind is to allow something like 
 `--style=style1  --file=1.pdf --rotate=90 --scale=0.5 -- --style=style2 style1 --file=2.pdf --  --pages style1 1-2 style2 z  style1 --rotate=0 4 --` 
instead of 

  `--pages  --file=1.pdf --rotate=90 --scale=0.5 1-2 --file=2.pdf --rotate=90 --scale=0.5 3 --file=1.pdf  --rotate=0 --scale=0.5 4 --`.

**Comment by jberkenbilt** [comment]
#1118 contains the first batch of CLI enhancements: non-positional options inside `--pages`, `--overlay`, and `--underlay`, and allowing `--overlay` and `--underlay` to be specified multiple times for stacking. Next I am going to start working on `QPDFAssember` and `QPDFSplitter`. I probably should have held 11.8.0 for this. I'm not planning on releasing 11.9.0 just for these changes.

**Comment by m-holger** [comment]
Low priority CLI enhancement:

remove requirement to terminate the `--pages` option with `--`

Instead of requiring `--` allow the option to be terminated by any other option that is not permitted inside the pages option (e.g. 
`--replace-input`). This makes the CLI less verbose, easier to use and eliminates one source of user error. The error messages produced in this situation are not helpful to users and likely to point them in the wrong direction (see for example #1183). The same enhancement applies to other options that  reqire termination with `--`.

**Comment by m-holger** [comment]
Medium priority CLI enhancement:

Allow input files to be specified outside of the `--pages` option.

Permit something like
```
--input --empty input1.pdf input2.pdf --password=marvin input3.pdf --pages --file-id=1 3-5 --file-id=3 1 file-id=2 z-6 --
```

This option would be particularly useful where qpdf is used for complex assembly of pages from two or more input files and where therefore the filenames are repeated multiple times inside the pages option (which suggests that there should be a short-form for eg `--file-id=1` such as `-i=1`). 

The second case where the option would be useful is where qpdf is used by another program using JSON. In this case it is likely that the user program already has a table of input files. Having to write out the filename for each page range adds unnecessary overhead both on the user and qpdf side.

**Comment by jberkenbilt** [comment]
For a variety of reasons, I don't see having time in the foreseeable future to finish this effort. I am going to park it for now. If anyone would like to take this over with my advice and review, let me know. I'm not married to this approach and actually find my proposed API to be a little cumbersome. I also think the existing helper API could use some work.

**Comment by m-holger** [comment]
I have added a page with the planned next steps for the 'pages epic' to the [wiki](https://github.com/qpdf/qpdf/wiki/Pages-epic-%3A-next-steps)

**Reply by qooxzuub** [reply]
I am interested in [Step 2b][1] (preserving named destinations when merging) of the pages epic. I have a pikepdf implementation for something like this. It basically

1. Builds a mapping of source-page references to their new locations in the output document.

2. Iterates the `/Names` tree, page `/Annots`, and document `/Outlines` to remap `GoTo` actions and destinations.

3. Handles name collisions by prefixing destinations (using a source-file ID and instance count) so that multiple merged documents don't overwrite each other's internal targets.

A few questions:

* Can 2b be implemented as a standalone enhancement to the existing merging logic in qpdf, or is it strictly dependent on the `QPDFAssembler` refactor in Step 1?

* Has any detailed design/spec work been done for step 2b, e.g., deciding the preferred strategy for name collisions in qpdf?

I would like to help flesh out the design and, if the approach is welcome, take a crack at the implementation in qpdf.

[1]: https://github.com/qpdf/qpdf/wiki/Pages-epic-%3A-next-steps#step-2b-add-option-to-preserve-named-destinations-when-merging-files

**Comment by rubyFeedback** [comment]
I don't have much to contribute to the proposal itself, as it is not quite within my own "reach" or level of expertise; I mostly use qpdf for simple stuff - query title of a pdf page, remove page number 55, merge two .pdf files. These kind of things. But I did want to say that I think I am probably not the only one who is curious about updates here. While such issue discussions are perfectly fine, I think it may be nice if qpdf as a project could give some updates every now and then. As that takes time, perhaps it could be done only a few times, say, twice a year? 

Something like a blog post but a bit more detailed than, say, the regular changelog. Something that can keep a perspective both into the future and the past, while being specific to the set goal (e. g. in the context here, of "pages epic" - by the way, I think I may also not be the only one who did not understand the name initially).
